### PR TITLE
Remove less individual sync calls to get compilation options.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/72930.

before, we make nearly 400 calls to get compilation options:

image

After, it's just 8:

image

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 1]
```diff
@@ -243,14 +243,24 @@ static SourceGeneratorExecutionVersionMap FilterToProjectCone(SourceGeneratorExe
...
                 CancellationToken cancellationToken)
             {
-                // Note: it's common to see a whole lot of project-infos change.  So attempt to collect that in one go
-                // if we can.
-                using var _ = PooledHashSet<Checksum>.GetInstance(out var projectInfoChecksums);
-                foreach (var (projectId, newProjectChecksums) in newProjectIdToStateChecksums)
-                    projectInfoChecksums.Add(newProjectChecksums.Info);
+                // Note: it's common to need to collect a large set of project-attributes and compilation options.  So
+                // attempt to collect all of those in a single call for each kind instead of a call for each instance
+                // needed.
```
Comment (ToddGrun) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:248]: This code only really needs to execute on changed projects, rignt? If so, could we move this code down there and have it only work on checksums from those projects?

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:248]: changed *and* added projects.  which is precisely what is in newProjectIdToStateChecksums.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**